### ISSUE TITLE
 pre-trained Caduceus model to the Huggingface hub，gradient

### ISSUE THREAD

**Issue by github2657529567** [kuleshov-group/caduceus#56]
We used the model on Huggingface, ps and ph, but there was a problem with the gradient, the gradient was very small, looking at the code we found that there was a problem with the initialisation of the model, referring to https://github.com/kuleshov-group/caduceus/issues/37， but still the gradient was very small. 

`import os
from transformers import AutoConfig, AutoTokenizer, AutoModelForSequenceClassification
import numpy as np
import torch
from sklearn.metrics import precision_recall_fscore_support
from transformers import DefaultDataCollator, Trainer, TrainingArguments

#os.environ['CUDA_VISIBLE_DEVICES'] = '0,1,2,3,4,5,6,7'
tokenizer = AutoTokenizer.from_pretrained(
    "/pf9550-bdp-A800/yehuicheng/h100/kuleshov-group/caduceus-ps_seqlen-131k_d_model-256_n_layer-16", trust_remote_code=True
)
config = AutoConfig.from_pretrained(
    "/pf9550-bdp-A800/yehuicheng/h100/kuleshov-group/caduceus-ps_seqlen-131k_d_model-256_n_layer-16",
    trust_remote_code=True,
    num_labels=12,
)
model = AutoModelForSequenceClassification.from_pretrained(
    "/pf9550-bdp-A800/yehuicheng/h100/kuleshov-group/caduceus-ps_seqlen-131k_d_model-256_n_layer-16",
    trust_remote_code=True,
    config=config,
    ignore_mismatched_sizes=True,
)

def print_grad(grad):
    print(grad)

# for name, param in model.named_parameters():
#     param.register_hook(lambda grad, name=name: print(f'{name} grad: {grad.norm()}'))

dataset_path="/pf9550-bdp-A800/yehuicheng/caduceus_env_ps/dataset/substrain_hcov_cls12_appendix"
from datasets import load_from_disk
trainset=load_from_disk(f"{dataset_path}/trainset")
evalset=load_from_disk( f"{dataset_path}/evalset")
testset=load_from_disk( f"{dataset_path}/testset")
def save_model(model, output_dir):
    model_save_path = os.path.join(output_dir, "trained_model.pth")
    torch.save(model.state_dict(), model_save_path)
    #print(f"model saved to ": {model_save_path}")

def evaluate_model(trainer, test_dataset):
    print("bengin evaluate testdataset")
    results = trainer.evaluate(test_dataset)
    #print(f{"results}")
    return results
def compute_metrics(p):
    logits,labels= p
    pred=np.argmax(logits, axis=-1)
    precision, recall, fscore, support = precision_recall_fscore_support(labels, pred, average="weighted")
    return {"precision":precision,"recall":recall,"fscore":fscore}
datacollator = DefaultDataCollator()
     
training_args=TrainingArguments(
    output_dir="/pf9550-bdp-A800/yehuicheng/Caduceus_ps/loadingsample",
    evaluation_strategy="steps",
    #gradient_checkpointing=True,
    eval_steps=10,
    save_steps=10,
    save_total_limit=10,
    learning_rate=5e-3,
    lr_scheduler_type= "cosine",
    warmup_ratio = 0.1,
    weight_decay=0.1,
    num_train_epochs=5,
    gradient_accumulation_steps=1,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    neftune_noise_alpha=0,
    max_grad_norm=1,
    bf16=True,
    logging_steps =1,
    report_to="wandb",
    optim="adamw_torch",
    save_safetensors=False
)

trainer=Trainer(
    model=model,
    args=training_args,
    train_dataset= trainset,
    eval_dataset= evalset,
    data_collator=datacollator,
    compute_metrics=compute_metrics
)

trainer.train()
save_model(model, training_args.output_dir)

evaluate_model(trainer, testset)`
![image](https://github.com/user-attachments/assets/6a6a3b45-a2de-4210-bacd-0da3d038d661)
![image](https://github.com/user-attachments/assets/a7f71eb0-1bfd-4f58-a16c-e5ef7bb01b81)

**Comment by yair-schiff** [comment]
I actually haven't done this sort of analysis myself for our models, so I am not sure if the grad norm ranges you're displaying are in line with my training. There does seem to be something weird going on with your loss curve though, I agree. Have you tried bigger batch size / playing around with the LR?

**Comment by leannmlindsey** [comment]
@github2657529567 

Did you use the most recent version of the the caduceus code ? check that it includes the fix [here](https://github.com/kuleshov-group/caduceus/blob/f05031fb760afda342d95b926a943b37dcdde97e/caduceus/modeling_caduceus.py#L500-L505):

Also, in our experiments in this [paper](https://www.biorxiv.org/content/10.1101/2024.09.09.612081v2): we observed significant sensitivity to batch size, learning rate and initialization weights in the state space models. These are also sensitive to any changes in both sequence length and model depth . You can see a table of the results of our experiments in [Supplementary Note 8](https://www.biorxiv.org/content/10.1101/2024.09.09.612081v2.supplementary-material).

My advice for doing your own training is to set up initial experiments using 4 random seeds, and a range of batch sizes and learning rates.  Train for a short amount of time to verify that your gradient is behaving as you wish, and then use those hyperparameters for further training.  This method worked for us as we continued further experiments.
